README.md

[![GitHub issues](https://img.shields.io/github/issues/DCL777/sensors2mqtt)](https://github.com/DCL777/sensors2mqtt/issues)
[![GitHub stars](https://img.shields.io/github/stars/DCL777/sensors2mqtt)](https://github.com/DCL777/sensors2mqtt/stargazers)
[![GitHub license](https://img.shields.io/github/license/DCL777/sensors2mqtt)](https://github.com/DCL777/sensors2mqtt/blob/master/LICENSE)
![GitHub top language](https://img.shields.io/github/languages/top/DCL777/sensors2mqtt)

# sensors2mqtt
Universal Sensor data transporter over MQTT.  Sensor like water flow sensor (=pulse-counter), DS18B20 , INA219,...  

For each sensor, an update can be choosen between:  10 (sec.) ,30,60, 300 (5min) , 900 (15min), 3600 (1h) , 10800 (3h), 21600 (6h) , 43200 (12h)

## Supported sensors:
- **Maximintegrated_DS18B20** temperature sensor
- **Generic_PulseCounter**
  - as a water-flow sensor
  - as gas meter counter
  - fully configurable for your application / sensor
- **TexasInstruments_INA219_4_20mA** 
  - as citern water height sensor. With a 'Submersible Water Level Transducer Sensor' 0-5m H2O  
  - fully configurable for your application
  - full range = 320mV, any current can be measured when the shunt resistor changes (not limited to 4-20mA)
- **TexasInstruments_INA226_4_20mA** 
  - as citern water height sensor. With a 'Submersible Water Level Transducer Sensor' 0-5m H2O  
  - fully configurable for your application
  - full range = 80mV, any current can be measured when the shunt resistor changes (not limited to 4-20mA)
- **Generic_SystemInfo**
  This sensor will send the following host system information:
   - Platform information (static)
   - CPU usage (dynamic)
   - Memory usage (dynamic)
   - Disk usage (dynamic) 
- **Generic_Ultrasonic**
  - This is generic PING-MODE ultrasonic sensor.  This uses a 'Trigger' and 'Echo'-signal
  - Supports averaging 
  - Example Devices:  JSN-SR04T, JSN-SR04T-2.0, JSN-SR04T-3.0, HC-SR04, HY-SRF05, US-100, ...  
- **Plantower_PMS5003**
  This is a Digital universal particle concentration sensor
  Measures: 
   - PM1.0, PM2.5 & PM10  value for CF=1，standard particle & under atmospheric environment
   - The number of particles with diameter 
      - beyond 0.3 um in 0.1 L of air.
      - beyond 0.5 um in 0.1 L of air.
      - beyond 1.0 um in 0.1 L of air.
      - beyond 2.5 um in 0.1 L of air.
      - beyond 5.0 um in 0.1 L of air.
      - beyond 10  um in 0.1 L of air.

## How to add more sensors? 
- copy and rename an existing sensor
- change the directory to ```[manufacturer]_[partnumber]```
- change the file name & the class name
    - ```[platform]_[manufacturer]_[partnumber]```   
    - or ```[platform]_Generic_[function]```
- change the content for your function
- The filename & class name must be exact (case sensitive) equal
- use the class name in 'sensors_settings.yaml' to activate the new module (dynamic loaded)
- [platform]
    - 'LINUX' - should work on all linux platforms
    - 'RPI'   - should work on all Raspberry PI platforms
    - 'ESP32' - should work on all ESP32 platforms
    - ...
- The 'platform' must be a YAML parameter for your sensor


# Config your sensors & MQTT broker info
- in directory ```~/sensors2mqtt/sensors2mqtt```
- copy sensors_settings.yaml to the same directory with a differend name: example: ```settings.yaml```
- Add or remove sensors in the new file.  Only sensors define here will be loaded at runtime
- Start with argument: -c option: ```./sensors2mqtt.py -c settings.yaml```
- Start with argument: -l option: ```./sensors2mqtt.py -c settings.yaml -l debug``` to get debug loggings
- To simplify starting:  go to ```~/sensors2mqtt``` 
  - and run: ```start_normal``` to have normal CLI output
  - and run: ```start_debug``` to have extra info in the CLI output 


# Setup
## on a Raspberry pi
Configure your rapsberry pi where the sensors are attachted:
- download the latest version of 'Raspberry Pi OS (32-bit) Lite' here: https://www.raspberrypi.org/downloads/raspberry-pi-os/
- enable SSH by adding a file named 'SSH' into the boot location
- if 1-wire is required (for DS18B20)
  - enable one-wire in config.txt (on your SD-Card) => with the correct pin number  (or in linux: ```sudo nano /boot/config.txt```)
  - ```dtoverlay=w1-gpio,gpiopin=22```
- If I2C is needed:
  - enable I2C in config.txt (on your SD-Card)
  - ```dtparam=i2c_arm=on```

   
- Start the Raspberry Pi with the prepared SD-Card
- connect over SSH | login
   - username: pi
   - password: raspberry
   - change your password with: 'passwd'

Download & Run: ```./single_file_setup```    
or execute the commands manually:

```bash
=> sudo apt update
=> sudo apt upgrade
=> sudo apt install git
=> sudo apt-get install i2c-tools    => for I2C only (like INA219)
=> sudo apt-get install python-smbus => for I2C only (like INA219)
=> sudo adduser pi i2c               => for I2C only (like INA219)
=> sudo reboot                       => for I2C only (like INA219)
=> sudo apt install python3-rpi.gpio
=> sudo apt-get install python3 
=> sudo apt-get install python3-pip
=> git clone https://github.com/DCL777/sensors2mqtt.git
=> cd sensors2mqtt
=> chmod +x ./start_normal
=> chmod +x ./start_normal
=> sudo pip3 install -r requirements.txt
=> chmod +x ./sensors2mqtt/sensors2mqtt.py
=> ./start_normal
```

## Use with home-assistant
Update your sensors section in configuration.yml with the new mqtt topics, for example:
```
- platform: mqtt
  state_topic: "Garage/Outside/Temperature"
  name: "Garage Temperature"
  unit_of_measurement: "°C"
- platform: mqtt
  state_topic: "Garage/Cistern/Temperature "
  name: "Cistern Temperature"
  unit_of_measurement: "°C"
```

# Hardware setup:
![screenshot](docs/images/hw.png?raw=true)

some hardware comments:
- DS18B20: needs a pull up resistor of 4k7
- my water flow puls counter works at 5VDC.  RPI needs 3V3 GPIO: so a resistor divider is used to lower the voltages
- INA219: change the shunt resistor to 10 ohm to have a good measurement
- Boost converter will convert the 5VDC to 12.5VDC needed for the 'Submersible Water Level Transducer Sensor'

# Example cli output: if level=logging.DEBUG
```
pi@regenputten:~/sensors2mqtt $ ./sensors2mqtt.py -c sensors_test.yaml
============================
Start Sensors to MQTT V0.0.1             2020-08-15 Dries Claerbout
============================
Start MQTT: Broker: 192.168.1.10 Username: ha

Sensors found
----------------------------------------------------------------------
 -> DS18B20      Maximintegrated         Temperature     1-Wire
    -> Sensor 1:
        -> path = /sys/bus/w1/devices/28-011939f0e24a/w1_slave
        -> mqtt_sub_dir = cistern
 -> PulseCounter         Generic         PulseCounter    GPIO
    -> Sensor 1:
        -> pin_number = 23
        -> mqtt_sub_dir = garden
        -> function = water-flow
    -> Sensor 2:
        -> pin_number = 24
        -> mqtt_sub_dir = home
        -> function = water-flow
 -> INA219 4-20mA        TexasInstruments        Current         I2C
    -> Sensor 1:
        -> address = 64
        -> mqtt_sub_dir = citern
        -> channel = 1
        -> calibration_low_mA = 3.96
        -> calibration_high_mA = 18.5
        -> full_range_value = 5000
        -> unit = mm
        -> mqtt_function_name = height
----------------------------------------------------------------------


2020-08-30 17:01:06,583 - DEBUG -
Starting main loop...
2020-08-30 17:01:06,585 - DEBUG - 10 seconds event 1598803266.5848727
2020-08-30 17:01:07,446 - INFO -     MQTT: garage/cistern/Temperature  20.937
2020-08-30 17:01:07,448 - INFO -     MQTT: garage/garden/water-flow/delta  0
2020-08-30 17:01:07,450 - INFO -     MQTT: garage/garden/water-flow/total  89
2020-08-30 17:01:07,453 - INFO -     MQTT: garage/home/water-flow/delta  0
2020-08-30 17:01:07,454 - INFO -     MQTT: garage/home/water-flow/total  15
2020-08-30 17:01:07,760 - INFO -     MQTT: garage/citern/height  19.567
2020-08-30 17:01:07,764 - INFO -     MQTT: garage/citern/bus-voltage  12.517
2020-08-30 17:01:10,002 - DEBUG - 10 seconds event 1598803270.0023887
2020-08-30 17:01:10,886 - INFO -     MQTT: garage/cistern/Temperature  21.0
2020-08-30 17:01:10,888 - INFO -     MQTT: garage/garden/water-flow/delta  0
2020-08-30 17:01:10,890 - INFO -     MQTT: garage/garden/water-flow/total  89
2020-08-30 17:01:10,891 - INFO -     MQTT: garage/home/water-flow/delta  0
2020-08-30 17:01:10,893 - INFO -     MQTT: garage/home/water-flow/total  15
2020-08-30 17:01:11,199 - INFO -     MQTT: garage/citern/height  19.567
2020-08-30 17:01:11,202 - INFO -     MQTT: garage/citern/bus-voltage  12.517
2020-08-30 17:01:20,009 - DEBUG - 10 seconds event 1598803280.0089643
2020-08-30 17:01:20,886 - INFO -     MQTT: garage/cistern/Temperature  20.937
2020-08-30 17:01:20,888 - INFO -     MQTT: garage/garden/water-flow/delta  0
2020-08-30 17:01:20,890 - INFO -     MQTT: garage/garden/water-flow/total  89
2020-08-30 17:01:20,892 - INFO -     MQTT: garage/home/water-flow/delta  0
2020-08-30 17:01:20,893 - INFO -     MQTT: garage/home/water-flow/total  15
2020-08-30 17:01:21,199 - INFO -     MQTT: garage/citern/height  19.567
2020-08-30 17:01:21,202 - INFO -     MQTT: garage/citern/bus-voltage  12.517
```
# Start at boot-time  (tested on raspberry-pi-os)

1. create a configuration file
```
sudo nano /lib/systemd/system/sensors2mqtt.service
```

2. set the content to: 
```
[Unit]
Description=Sensors to MQTT Service
After=multi-user.target


[Service]
Type=idle
WorkingDirectory=/home/pi/sensors2mqtt/sensors2mqtt
ExecStart=/usr/bin/python3 /home/pi/sensors2mqtt/sensors2mqtt/sensors2mqtt.py -c settings.yaml

[Install]
WantedBy=multi-user.target

```
3. set ter permission to 644:
```
sudo chmod 644 /lib/systemd/system/sensors2mqtt.service
```
4. enable the service
```
sudo systemctl daemon-reload
sudo systemctl enable sensors2mqtt.service
```
5. reboot
```
sudo reboot
```

6. check the status for the service
```
sudo systemctl status sensors2mqtt.service
```
or 
```
ps aux | grep sensor
```

# Change RPI fixed IP-address

```
sudo nano /etc/dhcpcd.conf
```

look for the following section and change it:
```
interface eth0
static ip_address=192.168.1.12/24
static routers=192.168.1.1
static domain_name_servers=192.168.1.1
```

# Special thanks to 
- https://pypi.org/project/smbus2/
- https://github.com/engonzal/DS18B20-mqtt


HISTORY.md
# History

- LINUX_Generic_DahuaEvents changed to LINUX_Dahua_Events
## V0.0.1  2022-07-20
- Breaking Change: Changed Python source directory from ./Sensors2mqtt to ./Sensors2mqtt/Sensors2mqtt
  Action required: update the paths in:  /lib/systemd/system/sensors2mqtt.service see readme for the updated data
- Breaking Change: Additional package required.  run the command below again
  sudo pip3 install -r requirements.txt  
- Generic_SystemInfo: Fixed issue with python 3.8 and higher   (distro.like() deprecated)
- Generic_Ultrasonic: 
    - Fixed some running issues
	- Added Timeout, so it's not crashing when not connected
 

sensors2mqtt/sensors2mqtt.py
#!/usr/bin/python3

#  This file is part of the sensors2mqtt distribution (https://github.com/DCL777/sensors2mqtt.git).
#  Copyright (c) 2020 Dries Claerbout
#
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, version 3.
#
#  This program is distributed in the hope that it will be useful, but
#  WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
#  General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program. If not, see <http://www.gnu.org/licenses/>.

# extra feature: log to file:
# http://blog.scphillips.com/posts/2013/07/getting-a-python-script-to-run-in-the-background-as-a-service-on-boot/


# Import package
#import os
#import glob
import time
import sys
import logging
import importlib
import argparse
import traceback
import paho.mqtt.client as mqtt
#import paho.mqtt.publish as publishpython
import yaml

#import platform
#from Sensor import Sensor
from Sensor import BColors


__version__ = "0.0.1"
__date__ = "2020-08-15"
__author__ = "Dries Claerbout"
__license__ = "AGPL-3.0+"
__status__ = "Debug"  # "Production"
#__name__ = "sensor2MQTT"


class sensors2mqtt():
    """Sensor2MQTT main class"""

    def __init__(self, config, logging):
        if config is None:
            self.config = "sensors_settings.yaml"
        else:
            self.config = config
        self.logging = logging
        self.my_sensor_list = []

    def load_config(self, path):
        """Load configurataion"""
        with open(path, 'r') as ymlfile:
            config = yaml.load(ymlfile, Loader=yaml.FullLoader)
        return config

    def init_mqtt(self, config_mqtt):
        """Initialize MQTT"""
        print(
            f"Start MQTT: Broker: {config_mqtt['broker']} Username: {config_mqtt['username']} \n")  # Password: {config_mqtt['password']}")

        try:
            client = mqtt.Client()
            client.username_pw_set(
                config_mqtt['username'], config_mqtt['password'])
            client.connect(config_mqtt['broker'])
            client.loop_start()
            return client
        except:
            print(f"{BColors.FAIL}Unable to start MQTT Client. {BColors.ENDC} Did you add the correct PASSWORD in your YAML file?\n")
            print("Possible reasons:")
            print("  -> wrong IP ADDRESS in your YAML file")
            print("  -> wrong PASSWORD in your YAML file")
            print("  -> is the MQTT server up and running?\n\n")
            sys.exit()

    def send_update(self, event):
        """Send Update"""
        logging.info(f"EVENT = {event}")
        #logging.info("EVENT = %s", event)
        for x in self.my_sensor_list:
            try:
                if int(x.get_update_interval()) <= int(event):
                    x.send_value_over_mqtt()
            except Exception as e:
                logging.error("\n\n\n%s\n" , traceback.format_exc() )

    def run(self):
        """RUN IT"""
        log_formatter = '%(asctime)s - %(levelname)s - %(message)s'
        if self.logging is None:
            # change this to  DEBUG, TEST, WARNING or ERROR to see more or less info
            logging.basicConfig(format=log_formatter, level=logging.ERROR)
        elif self.logging == 'debug':
            # change this to  DEBUG, TEST, WARNING or ERROR to see more or less info
            logging.basicConfig(format=log_formatter, level=logging.DEBUG)
        else:
            # change this to  DEBUG, TEST, WARNING or ERROR to see more or less info
            logging.basicConfig(format=log_formatter, level=logging.ERROR)
        logger = logging.getLogger(__name__)

        print(f"{BColors.HEADER}============================")
        print(
            f"Start Sensors to MQTT V{__version__}             {__date__} {__author__}")
        print(f"============================{BColors.ENDC}")

        config_yaml = self.load_config(self.config)
        config_mqtt = config_yaml['mqtt']
        client = self.init_mqtt(config_mqtt)

        mqtt_top_dir_name = config_mqtt['top_dir_name']

        self.my_sensor_list = []
        config_sensors = config_yaml['Sensors']

        logger.debug("Loading modules:")
        logger.debug("================")

        if config_sensors is None:
            print(
                f"\n{BColors.FAIL}No Sensors configured in the settings-file: {self.config}{BColors.ENDC}")
            print(f"\nNothing to do, so this program will be stopped\n")
            exit()

        # load only the sensors needed
        for a_sensor_class in config_sensors:
            #print(f"a_sensor_class =  {a_sensor_class}")
            sensor_list = config_sensors.get(f"{a_sensor_class}")
            for a_sensor in sensor_list:
                #print(f"a_sensor =  {a_sensor}")
                #print(f"platform =  {a_sensor['platform']}")
                # https://realpython.com/python-import/
                class_to_load = f"sensors.{a_sensor_class}.{a_sensor['platform']}_{a_sensor_class}"
                #print(f"class_to_load =  {class_to_load}")
                module = importlib.import_module(class_to_load)
                logger.debug(f"    --> CLASS: {class_to_load}")
                a_class = getattr(
                    module, f"{a_sensor['platform']}_{a_sensor_class}")
                my_class = a_class(client, a_sensor, mqtt_top_dir_name)
                self.my_sensor_list.append(my_class)

        lowest_update_interval = 43200  # = 12h

        print("\nSensors found")
        print("----------------------------------------------------------------------")
        for x in self.my_sensor_list:
            x.print_info()
            if x.get_update_interval() < lowest_update_interval:
                lowest_update_interval = x.get_update_interval()
            print(
                "----------------------------------------------------------------------")
        print(f"Lowest Update Interval: {lowest_update_interval} seconds")

        print("\n\n\nStarting main loop...")
        self.send_update(43200)
        time.sleep(int(10) - ((time.time()) % float(10)))

        while True:
            try:
                event = 43200
# seconds  10,30,60, 300 (5min) , 900 (15min), 3600 (1h) , 10800 (3h), 21600 (6h) , 43200 (12h)
                if (time.time() % 43200.0) < 2:
                    event = 43200
                elif (time.time() % 21600.0) < 2:
                    event = 21600
                elif (time.time() % 10800.0) < 2:
                    event = 10800
                elif (time.time() % 3600.0) < 2:
                    event = 3600
                elif (time.time() % 900.0) < 2:
                    event = 900
                elif (time.time() % 300) < 2:
                    event = 300
                elif (time.time() % 60) < 2:
                    event = 60
                elif (time.time() % 30) < 2:
                    event = 30
                elif (time.time() % 10) < 2:
                    event = 10
                # else:
                #  logging.error("\n  ===========> event not in range (!) check" )

                self.send_update(event)

                time.sleep(int(lowest_update_interval) -
                           ((time.time()) % float(lowest_update_interval)))
            except KeyboardInterrupt:
                print('\ncaught keyboard interrupt!, bye')
                for x in self.my_sensor_list:
                    x.on_exit()
                sys.exit()


parser = argparse.ArgumentParser()
parser.add_argument('-c', dest='config', default="sensors_settings.yaml", help='path to your config file i.e. sensors_settings.yaml (= default value)')
parser.add_argument('-l', dest='logging', default="error", help='debug or error')
args = parser.parse_args()
sens2mqtt = sensors2mqtt(args.config, args.logging)
sens2mqtt.run()
